test/performance/odp_stash_perf.c

```diff
@@ -0,0 +1,451 @@
+			break;
+
+		switch (opt) {
+		case 'c':
+			test_options->num_cpu = atoi(optarg);
+			break;
+		case 'n':
+			test_options->num_stash = atoi(optarg);
+			break;
+		case 'b':
+			test_options->max_burst = atoi(optarg);
+			break;
+		case 'r':
+			test_options->num_round = atoi(optarg);
+			break;
+		case 's':
+			test_options->strict = 1;
+			break;
+		case 'h':
+			/* fall through */
+		default:
+			print_usage();
+			ret = -1;
+			break;
+		}
+	}
+
+	if (test_options->num_stash > MAX_STASHS) {
+		printf("Too many stashes %u. Test maximum %u.\n",
+		       test_options->num_stash, MAX_STASHS);
+		return -1;
+	}
+
+	return ret;
+}
+
+static int create_stashes(test_global_t *global)
+{
+	uint32_t i;
+	test_options_t *test_options = &global->options;
+
+	uint32_t num_stash = test_options->num_stash;
+	uint32_t num_round = test_options->num_round;
+	odp_stash_t *stash = global->stash;
+
+	printf("\nTesting %s stashs\n",
+	       test_options->strict == 0 ? "NORMAL" : "STRICT_SIZE");
+	printf("  num rounds           %u\n", num_round);
+	printf("  num stashs           %u\n", num_stash);
+	printf("  max burst size       %u\n", test_options->max_burst);
+
+	for (i = 0; i < num_stash; i++) {
+		odp_stash_param_t stash_param;
+
+		odp_stash_param_init(&stash_param);
+		stash_param.num_obj    = 1000;
+		stash_param.obj_size   = sizeof(uint64_t);
+		stash_param.strict_size = test_options->strict;
+
+		stash[i] = odp_stash_create("test_stash_u64", &stash_param);
+		if (stash[i] == ODP_STASH_INVALID) {
+			ODPH_ERR("Stash create failed\n");
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+static int destroy_stashes(test_global_t *global)
+{
+	odp_stash_t *stash = global->stash;
+	test_options_t *test_options = &global->options;
+	uint32_t num_stash = test_options->num_stash;
+
+	for (uint32_t i = 0; i < num_stash; i++) {
+		if (odp_stash_destroy(stash[i])) {
+			ODPH_ERR("Stash destroy failed\n");
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+static int run_test(void *arg)
+{
+	uint64_t c1, c2, cycles, nsec;
+	odp_time_t t1, t2;
+	uint32_t rounds;
+	int num_stored;
+	test_stat_t *stat;
+	test_global_t *global = arg;
+	test_options_t *test_options = &global->options;
+	odp_stash_t stash;
+	uint64_t tmp[32] = {0};
+	uint64_t num_retry = 0;
+	uint64_t ops = 0;
+	uint32_t num_stash = test_options->num_stash;
+	uint32_t num_round = test_options->num_round;
+	int thr = odp_thread_id();
+	int ret = 0;
+	uint32_t i = 0;
+	uint32_t max_burst = test_options->max_burst;
+
+	stat = &global->stat[thr];
+
+	/* Start all workers at the same time */
+	odp_barrier_wait(&global->barrier);
+
+	t1 = odp_time_local();
+	c1 = odp_cpu_cycles();
+
+	for (rounds = 0; rounds < num_round; rounds++) {
+		do {
+			stash = global->stash[i++];
+
+			if (i == num_stash)
+				i = 0;
+
+			num_stored = odp_stash_put_u64(stash, tmp, max_burst);
+
+			if (num_stored <= 0)
+				num_retry++;
+		} while (num_stored <= 0);
+
+		if (odp_stash_get_u64(stash, tmp, max_burst) != num_stored) {
```

Comment:
Depending on thread timing, `odp_stash_get_u64()` could return < num_stored objects and this would still be OK according to the spec.